lib/geo/wkb/encoder.ex

```diff
@@ -18,238 +36,169 @@ defmodule Geo.WKB.Encoder do
-        acc = encode_coordinates(acc, %LineStringZ{coordinates: line})
-        {nil, acc}
-      end)
-
-    writer
-  end
-
-  defp encode_coordinates(writer, %MultiPoint{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%Point{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiPointZ{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%PointZ{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiLineString{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%LineString{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiLineStringZ{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%LineStringZ{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiPolygon{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%Polygon{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiPolygonZ{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%PolygonZ{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
+  defp add_srid(type), do: type + @wkbsridflag
+
+  def encode!(geom, endian \\ :ndr)
+
+  for {endian, endian_atom, modifier} <- [{1, :ndr, quote(do: little)}, {0, :xdr, quote(do: big)}] do
+    def encode!(geom, unquote(endian_atom)) do
+      {type, rest} = do_encode(geom, unquote(endian_atom))
+
+      binary =
+        if geom.srid do
+          <<add_srid(type)::unquote(modifier)-32, geom.srid::unquote(modifier)-32>>
+        else
+          <<type::unquote(modifier)-32>>
+        end
+
+      IO.iodata_to_binary([unquote(endian), binary, rest])
```

Comment:
Consider removing this IO.iodata_to_binary call:

```suggestion
      [unquote(endian), binary, rest]
```

otherwise, `Geo.WKB.dump_iodata` always returns a _binary_! (which is totally fine, but often times we'd happily work with the unchanged iodata instead for maximum performance)

Reply:
Good point! I will update that